libraries/protocol/include/graphene/protocol/custom_authority.hpp

```diff
@@ -0,0 +1,136 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#pragma once
+#include <graphene/protocol/base.hpp>
+#include <graphene/protocol/asset.hpp>
+#include <graphene/protocol/authority.hpp>
+#include <graphene/protocol/restriction.hpp>
+
+namespace graphene { namespace protocol {
+
+   /**
+    * @brief Create a new custom authority
+    * @ingroup operations
+    */
+   struct custom_authority_create_operation : public base_operation {
+      struct fee_parameters_type {
+         uint64_t basic_fee = GRAPHENE_BLOCKCHAIN_PRECISION;
+         uint32_t price_per_byte = GRAPHENE_BLOCKCHAIN_PRECISION / 10;
+      };
+
+      /// Operation fee
+      asset fee;
+      /// Account which is setting the custom authority; also pays the fee
+      account_id_type account;
+      /// Whether the custom authority is enabled or not
+      bool enabled;
+      /// Date when custom authority becomes active
+      time_point_sec valid_from;
+      /// Expiration date for custom authority
+      time_point_sec valid_to;
+      /// Tag of the operation this custom authority can authorize
+      unsigned_int operation_type;
+      /// Authentication requirements for the custom authority
+      authority auth;
+      /// Restrictions on operations this custom authority can authenticate
+      vector<restriction> restrictions;
+
+      extensions_type extensions;
+
+      account_id_type fee_payer()const { return account; }
+      void validate()const;
+      share_type calculate_fee(const fee_parameters_type& k)const;
+   };
+
+   /**
+    * @brief Update a custom authority
+    * @ingroup operations
+    */
+   struct custom_authority_update_operation : public base_operation {
+      struct fee_parameters_type {
+         uint64_t basic_fee = GRAPHENE_BLOCKCHAIN_PRECISION;
+         uint32_t price_per_byte = GRAPHENE_BLOCKCHAIN_PRECISION / 10;
+      };
+
+      /// Operation fee
+      asset fee;
+      /// Account which owns the custom authority to update; also pays the fee
+      account_id_type account;
+      /// ID of the custom authority to update
+      custom_authority_id_type authority_to_update;
+      /// Change to whether the custom authority is enabled or not
+      optional<bool> new_enabled;
+      /// Change to the custom authority begin date
+      optional<time_point_sec> new_valid_from;
+      /// Change to the custom authority expiration date
+      optional<time_point_sec> new_valid_to;
+      /// Change to the authentication for the custom authority
+      optional<authority> new_auth;
+      /// Set of indexes of restrictions to remove
+      flat_set<uint16_t> restrictions_to_remove;
```

Comment:
A user should not broadcast two update ops simultaneously, she should use a single update op that affects all appropriate restrictions.

Restrictions don't have expirations, only custom authorities do, and even when they expire, they aren't removed from the database immediately, they're only automatically removed during the blockchain maintenance period, so there's no race.

I think the index is a sufficient ID. It is stable and predictable by clients and cannot be changed except by user action.

Reply:
Although I didn't find a use case right now, I guess there would be some. It has the potentials. "A user" can be a service provider but not only an individual, this operation can be generated via APIs called by its clients, think further, this operation can even be in a proposal thus timing of approvals could matter, etc, so it would be easier to integrate if the user can specify the ID.

Reply:
We can also solve this issue later, after https://github.com/bitshares/bsips/issues/185 is ready.

Reply:
I've decided you're right, using the index as ID is too unpredictable. I am revising the code to give the restrictions stable IDs, which are assigned consecutively by the chain, but will never change once assigned. Thus restrictions on the object are now stored in a `flat_map<uint16_t, restriction>` rather than a simple `vector<restriction>`.

I think 65536 is a sufficient ID space; it would be unlikely (and expensive) for someone to create enough restrictions to cause a wraparound, but if they do, the chain will detect it and throw an exception.